lib/Reader/readerir.cpp

```diff
@@ -1418,16 +1418,75 @@ Type *GenIR::getClassType(CORINFO_CLASS_HANDLE ClassHandle, bool IsRefClass,
     // default comparator here.
     std::sort(DerivedFields.begin(), DerivedFields.end());
 
+    // If we find overlapping fields, we'll stash them here so we can look
+    // at them collectively.
+    std::vector<std::pair<uint32_t, Type *>> OverlappingFields;
+
     // Now walk the fields in increasing offset order, adding
     // them and padding to the struct as we go.
     for (const auto &FieldPair : DerivedFields) {
       const uint32_t FieldOffset = FieldPair.first;
       CORINFO_FIELD_HANDLE FieldHandle = FieldPair.second;
 
-      // Bail out for now if we see a union type.
+      // Prepare to add this field to the collection.
+      //
+      // If this field is a ref class reference, we don't need the full
+      // details on the referred-to class, and asking for the details here
+      // causes trouble with certain recursive type graphs, for example:
+      //
+      // class A { B b; }
+      // class B : extends A { int c; }
+      //
+      // We need to know the size of A before we can finish B. So we can't
+      // ask for B's details while filling out A.
+      CORINFO_CLASS_HANDLE FieldClassHandle;
+      CorInfoType CorInfoType = getFieldType(FieldHandle, &FieldClassHandle);
+      bool GetFieldDetails = (CorInfoType != CORINFO_TYPE_CLASS);
+      Type *FieldTy = getType(CorInfoType, FieldClassHandle, GetFieldDetails);
+      // If we don't get the details now, make sure to ask
+      // for them later.
+      if (!GetFieldDetails) {
+        DeferredDetailClasses.push_back(FieldClassHandle);
+      }
+
+      // If we see an overlapping field, we need to handle it specially.
       if (FieldOffset < ByteOffset) {
-        ASSERT(IsUnion);
-        throw NotYetImplementedException("union types");
+        assert(IsUnion && "unexpected overlapping fields");
+        // TypedByref and String get special treatement which we will skip
+        // in processing overlaps.
+        assert(!IsTypedByref && "No overlap expected in this type");
+        assert(!IsString && "No overlap expected in this type");
+        if (OverlappingFields.empty()) {
+          // The previously processed field is also part of the overlap
+          // set. Back it out of the main field collection and add it to the
+          // overlap collection instead.
+          Type *PreviousFieldTy = Fields.back();
+          Fields.pop_back();
+          uint32_t PreviousSize =
+              DataLayout->getTypeSizeInBits(PreviousFieldTy) / 8;
+          uint32_t PreviousOffset = ByteOffset - PreviousSize;
+          addFieldsRecursively(OverlappingFields, PreviousOffset,
+                               PreviousFieldTy);
+        }
+
+        // Add the current field to the overlap set.
+        uint32_t FieldSize = DataLayout->getTypeSizeInBits(FieldTy) / 8;
+        OverlappingFields.push_back(std::make_pair(FieldOffset, FieldTy));
```

Comment:
Should this be here?  If the field is a scalar, it looks like it's redundant with the subsequent call to `addFieldsRecursively`.  If the field is a struct, it seems odd that we'd want top-level struct fields on the list but not struct-typed sub-fields (which `addFieldsRecursively` will not include in the list)..